sbncode/TPCReco/VertexStub/StubBuilder.cxx

```diff
@@ -0,0 +1,232 @@
+
+  std::vector<unsigned> pfp_nhit(pfps.size(), 0);
+
+  for (const art::Ptr<recob::Hit> h: hits) {
+    for (unsigned i_pfp = 0; i_pfp < pfps.size(); i_pfp++) {
+      const std::vector<art::Ptr<recob::Hit>> &thispfpHits = pfp_hits[i_pfp];
+      bool found = false;
+      for (const art::Ptr<recob::Hit> &pfp_h: thispfpHits) {
+        if (h == pfp_h) {
+          found = true;
+          pfp_nhit[i_pfp] ++;
+          break;
+        }
+      }
+      if (found) break;
+    }
+  }
+
+  for (unsigned i_pfp = 0; i_pfp < pfps.size(); i_pfp++) {
+    if (pfp_nhit[i_pfp] >= hits.size() / 2) return i_pfp;
+  }
+
+  return -1;
+}
+
+std::vector<art::Ptr<recob::Hit>> CollectHits(
+    const std::vector<art::Ptr<recob::Hit>> &hits, 
+    const sbn::VertexHit &vhit, 
+    const recob::Hit &vhit_hit, 
+    const geo::Point_t &vertex, 
+    const geo::GeometryCore *geo, 
+    const detinfo::DetectorPropertiesData &dprop) {
+
+  // project the vertex onto the wireplane
+  float vert_wf = geo->WireCoordinate(vertex, vhit_hit.WireID());
+  float vert_x = vertex.x();
+
+  // get the vertex hit-coordinates
+  int hit_w = vhit_hit.WireID().Wire;
+  float hit_x = dprop.ConvertTicksToX(vhit_hit.PeakTime(), vhit_hit.WireID());
+  geo::PlaneID vhit_plane = vhit_hit.WireID();
+
+  // get the line-segment slope / intercept
+  float slope = (hit_x - vert_x) / (hit_w - vert_wf);
+  float intercept = hit_x - slope * hit_w;
+
+  // Include wires one after the hit and one before the vtx
+  int vert_w = (int)((hit_w > vert_wf) ? std::floor(vert_wf) : std::ceil(vert_wf));
+  hit_w = hit_w + ((hit_w > vert_wf) ? 1 : -1);
+
+  // get all the hits that overlap between these two points
+  std::vector<art::Ptr<recob::Hit>> ret;
+  for (art::Ptr<recob::Hit> h: hits) {
+    int this_hit_w = h->WireID().Wire;
+    geo::PlaneID this_hit_plane = h->WireID();
+    if (this_hit_plane == vhit_plane &&  // check plane
+       ((this_hit_w <= hit_w && this_hit_w >= vert_w) || // check overlap
+        (this_hit_w >= hit_w && this_hit_w <= vert_w))) {
+      float this_proj_x = slope * this_hit_w + intercept;
+
+      float h_time_lo = h->PeakTime() - h->RMS();
+      float h_x_A = dprop.ConvertTicksToX(h_time_lo, h->WireID());
+      float h_time_hi = h->PeakTime() + h->RMS();
+      float h_x_B = dprop.ConvertTicksToX(h_time_hi, h->WireID());
+
+      if ((h_x_A >= this_proj_x && h_x_B <= this_proj_x) ||
+          (h_x_A <= this_proj_x && h_x_B >= this_proj_x)) {
+        ret.push_back(h);
+      }
+    }
+  }
+  
+  return ret;
+}
+  
+
+void sbn::StubBuilder::Setup(const art::Event &e, const art::InputTag &pfplabel, const art::InputTag &trklabel) {
+  // Clear out old data
+  fSlicePFPHits.clear();
+  fSlicePFPs.clear();
+  fSliceHits.clear();
+  fSliceTrkHits.clear();
+  fSliceTrkTHMs.clear();
+  fSliceTrks.clear();
+
+  // Build the map of the slice key to the hits
+  art::Handle<std::vector<recob::Slice>> slice_handle;
+  e.getByLabel(pfplabel, slice_handle);
```

Comment:
I'd say that's a feature, not a bug. I don't know if it's always a good idea to crash if slices are missing. Just not producing any output (which is what this code does) I think is reasonable.

Reply:
Just a drive-by to drop in my 2c here.
I don't think "just don't produce any output if you can't find the slices" is a good idea. If the user meant to not specify valid slices, they could use a blank label (and it would be appropriate to have special handling for that).

Scientific software isn't an autopilot or a pacemaker, we have no obligation to make a best effort when there are problems. The much bigger risk is that we do months of production processing and then discover no stubs were made because of a trivial misconfiguration, which the code covered up.

Similarly for things that "should never happen". The right thing is to blow up the job with a verbose explanation of the problem so it can be understood and fixed properly, not sweep it under the rug. Not saying you're guilty of this, but I've definitely seen a lot of it in physics code and it drives me mad.

Reply:
> not [...] a pacemaker, we have no obligation to make a best effort when there are problems

I will definitely feel safe if you ever decide to go to medical.